R/simulate.R

```diff
@@ -0,0 +1,90 @@
+i2p_simulate <- function(dat, param, nsamples = NULL, type = "estimate") {
+
+  type <- match.arg(type, choices = c("estimate", "observations"))
+  files <- c("ab.stan", "gaussian_process.stan",
+             "observed_in_window.stan", "convolve.stan")
+
+  functions <-
+    paste0("\n functions{ \n", paste(purrr::map_chr(
+      files, ~paste(readLines(here::here("stan", "functions", .)),
+                    collapse = "\n")),
+      collapse = "\n"), "\n }")
+
+  ## fix rstan / cmdstanr compatibility issue
+  functions <- gsub("array\\[\\] ([a-z]+) ([a-z_]+)", "\\1[] \\2", functions)
+
+  rstan::expose_stan_functions(rstan::stanc(model_code = functions))
```

Comment:
obviously bit of a hack using the stan code but this is nice!

Reply:
After writing this function it occurred to me that `cmdstanr`'s [$generate_quantities](https://mc-stan.org/cmdstanr/reference/model-method-generate-quantities.html) could be a much clearer path to the same outcome but I'm not 100% sure (and haven't tried) whether it would be up to the task.

Reply:
Yes, I think so but it would have much the same effect. Not clear to me that you could do that without having a second gq version of the model. If you have to do that then some of the benefits start to get lost. Having it even partially in R is really nice for exploration/debugging/sanity. This is what keeps drawing me into Julia/Turing. I spoke to a few people who've been using it at scale and sadly apparently has some numerical stability issues at the moment.